test/test_framework.py

```diff
@@ -528,6 +528,40 @@ class NoneEvent(EventBase):
         with self.assertRaises(RuntimeError):
             pub.on_a.define_event("foo", MyFoo)
 
+    def test_event_key_roundtrip(self):
+        class MyEvent(EventBase):
+            pass
+
+        class MyNotifier(Object):
+            foo = Event(MyEvent)
+
+        class MyObserver(Object):
+            def __init__(self, parent, key):
+                super().__init__(parent, key)
+                self.seen = []
+
+            def on_foo(self, event):
+                self.seen.append(event.handle)
+
+        framework1 = self.create_framework()
+        pub1 = MyNotifier(framework1, "1")
+        obs1 = MyObserver(framework1, "1")
+        framework1.observe(pub1.foo, obs1)
+        pub1.foo.emit()
+
+        framework1.commit()
+        framework1.close()
+
+        framework2 = self.create_framework()
+        pub2 = MyNotifier(framework2, "2")
+        obs2 = MyObserver(framework2, "2")
+        framework2.observe(pub2.foo, obs2)
+        pub2.foo.emit()
+
+        handle1 = obs1.seen[0]
+        handle2 = obs2.seen[0]
+        self.assertNotEqual(handle1.key, handle2.key)
```

Comment:
I assume that this wouldn't hold true if we didn't save the events, as the key ends up holding the event count? Which would otherwise be reset to 0 in framework2?

Reply:
Yeah, this test fails if the rest of the change is not in place (I created the test first to confirm this).  That is, with the second framework (e.g., during the next hook invocation), if the same type of event is emitted but was also deferred (each with different contextual data), they could have the same handle since the count got reset between hooks.  Thus, the contextual data from one would get overwritten.